### PR TITLE
Experimental: xstate example

### PR DESCRIPTION
**Pull request by jackkleeman**

### CODE REVIEW

File: typescript/xstate/src/lib.ts [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+interface FakeParent<TLogic extends AnyStateMachine> extends AnyActorRef {
+  _send: (event: EventFromLogic<TLogic>) => void;
+}
+
+export async function createActor<TLogic extends AnyStateMachine>(ctx: restate.RpcContext, api: XStateApi<TLogic>, systemName: string, logic: TLogic extends AnyStateMachine ? AreAllImplementationsAssumedToBeProvided<TLogic['__TResolvedTypesMeta']> extends true ? TLogic : MissingImplementationsError<TLogic['__TResolvedTypesMeta']> : TLogic, options?: ActorOptions<TLogic>): Promise<Actor<TLogic>> {
+  const system = await createSystem(ctx, api, systemName)
+  const snapshot = await ctx.get<Snapshot<unknown>>("snapshot") ?? undefined;
+
+  const parent: FakeParent<TLogic> = {
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:222]: hack; I can't find a way to create an actor with a particular system set, so I had to create a 'fake' parent for the new actor which has a system already set

File: typescript/xstate/src/lib.ts [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    },
+    send: async (ctx: restate.RpcContext, systemName: string, request?: { scheduledEvent?: SerialisableScheduledEvent, source?: SerialisableActorRef, target?: SerialisableActorRef, event: AnyEventObject }): Promise<Snapshot<unknown> | undefined> => {
+      if (!request) {
+        throw new TerminalError("Must provide a request")
+      }
+
+      if (request.scheduledEvent) {
+        const events = await ctx.get<{ [key: string]: SerialisableScheduledEvent }>("events") ?? {}
+        const scheduledEventId = createScheduledEventId(request.scheduledEvent.source, request.scheduledEvent.id)
+        if (!(scheduledEventId in events)) {
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:296]: Restate doesn't yet support directly cancellable delayed calls, so I sort of implement it myself by just checking to see if the event is still scheduled, and also that it hasn't been replaced

File: typescript/xstate/src/lib.ts [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+  const system: RestateActorSystem<T> = {
+    ctx,
+    api,
+    systemName,
+
+    _bookId: () => ctx.rand.uuidv4(),
+    _register: (sessionId, actorRef) => {
+      if (actorRef.id in childrenByID) {
+        // rehydration case; ensure session ID maintains continuity
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:142]: general question; is my use of `sessionid` right? i treat it as a particular execution of an actor, so we want the id to be unique, but also deterministic on rehydrate (ie, on the next event where we start up the state machine all over again). thats why I deliberately store session ids in state, and then correlate them back with the rehydrated children using `id` in order to set the same session ids

File: typescript/xstate/src/lib.ts [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+      ctx.set("events", events)
+    },
+  }
+
+  const system: RestateActorSystem<T> = {
+    ctx,
+    api,
+    systemName,
+
+    _bookId: () => ctx.rand.uuidv4(),
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:139]: instead of persisting a counter just using deterministic uuids here

File: typescript/xstate/src/lib.ts [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+    },
+  }
+}
+
+export const bindXStateRouter = <
+  TLogic extends AnyStateMachine,
+>(server: restate.RestateServer, path: string, logic: TLogic): restate.RestateServer => {
+  return server
+    .bindKeyedRouter(path, restate.keyedRouter(actorMethods(path, logic)))
+    .bindRouter(`${path}.promises`, restate.router(promiseMethods(path, logic)))
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:341]: This is a separate service to offload long-running promise execution to, so that the main event loop isnt blocked (restate processes events in a keyed service in series). this is more of a restate hack right now, we would like a way to have non locking methods on the same service

File: typescript/xstate/src/lib.ts [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        delete events[scheduledEventId]
+        ctx.set("events", events)
+      }
+
+      const root = (await createActor(ctx, api, systemName, logic)).start();
+
+
+      let actor;
+      if (request.target) {
+        actor = (root.system as RestateActorSystem<any>).actor(request.target.sessionId)
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:313]: I defined this way to get a child by session id given that an event is for a particular session id, and we don't want it to go to something that has stopped and restarted

File: typescript/xstate/src/lib.ts [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+        if (!actor) {
+          throw new TerminalError(`Actor ${request.target.id} not found; it may have since stopped`)
+        }
+      } else {
+        actor = root
+      }
+
+      (root.system as RestateActorSystem<any>)._relay(request.source, actor, request.event)
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:321]: this is the core of the whole thing. get the actor, send the message, store the new persisted snapshot

File: typescript/xstate/src/promise.ts [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+      rs.ctx.send(rs.api.promise).invoke({
+        systemName: rs.systemName,
+        self: serialiseActorRef(self),
+        srcs: actorSrc(self),
+        input: state.input
+      });
+
+      // note that we sent off the promise so we don't do it again
+      (system as any)._relay(self, self, {
```
Comment (jackkleeman) [inline typescript/xstate/src/promise.ts:105]: This differs slightly from the built in fromPromise actor because i think there we just rely on in-memory state to keep track of the promise having already been run, whereas this needs to be resistant to rehydration, i guess? in general we may have a problem of re-executing some types of actions, I think

File: typescript/xstate/src/promise.ts [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+  if (actor === undefined) {
+    return []
+  }
+  if (typeof actor.src !== "string") {
+    return []
+  }
+  return [actor.src, ...actorSrc(actor._parent)]
+}
+
+export const promiseMethods = <TLogic extends AnyStateMachine>(path: string, logic: TLogic) => {
```
Comment (jackkleeman) [inline typescript/xstate/src/promise.ts:135]: this stuff all runs in a separate restate service that is unkeyed (ie non locking) but has access to the same state machine logic. the idea is that it gets called with all the information it needs to find the promise creating function in the state machine (could be in a deep chain of nested state machines) and also the inputs to call that function with. it then calls, awaits, and fires an event back to the main loop on completion

basically just using the restate async runtime instead of the JS async runtime

File: typescript/xstate/src/promise.ts [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+      };
+    },
+    getPersistedSnapshot: (snapshot) => snapshot,
+    restoreSnapshot: (snapshot: any) => snapshot
+  };
+
+  return logic;
+}
+
+function actorSrc(actor?: AnyActorRef): string[] {
```
Comment (jackkleeman) [inline typescript/xstate/src/promise.ts:125]: tries to determine the chain of `src` fields which is the best way i could find to derive a 'path' through the state machine definition that helps you find the promise creator

File: typescript/xstate/src/promise.ts [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+      self,
+      srcs,
+      input
+    }: { systemName: string, self: SerialisableActorRef, srcs: string[], input: NonReducibleUnknown }) => {
+      console.log("run promise with srcs", srcs, "in system", systemName, "with input", input)
+
+      const [promiseSrc, ...machineSrcs] = srcs
+
+      let stateMachine: AnyStateMachine = logic;
+      for (const src of machineSrcs) {
```
Comment (jackkleeman) [inline typescript/xstate/src/promise.ts:150]: apply the 'path' through the state machine definitions, checking for each one in either the implementations or directly in the defintiion

File: typescript/xstate/src/lib.ts [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+  const childrenByID = await ctx.get<{ [key: string]: SerialisableActorRef }>("children") ?? {}
+
+  let idCounter = 0;
+  const children = new Map<string, AnyActorRef>();
+  const keyedActors = new Map<keyof T['actors'], AnyActorRef | undefined>();
+  const reverseKeyedActors = new WeakMap<AnyActorRef, keyof T['actors']>();
+  const observers = new Set<Observer<InspectionEvent>>();
+
+  const scheduler = {
+    schedule(_source: AnyActorRef, _target: AnyActorRef, event: EventObject, delay: number, id: string | undefined): void {
```
Comment (jackkleeman) [inline typescript/xstate/src/lib.ts:84]: scheduler implementation is the first way we can do stuff across time; allows us to delay sending events by using restates `sendDelayed` abstraction. however this is not currently cancellable so we need to keep track of what we have sent and cancelled and check again on arrival
